File: web_ui/routes/detail.py
"""Memory detail view route."""

import json
import time
from urllib.parse import quote

from starlette.requests import Request
from starlette.responses import HTMLResponse, RedirectResponse
from starlette.routing import Route

from memory.licence import (
    LICENCE_CHOICES,
    LICENCE_LABELS,
    classify_memories,
    is_classifiable_key,
    note_for_reclassification,
    resolve_licence,
    validate_licence_note,
)
from memory.provenance import (
    PROVENANCE_CHOICES,
    PROVENANCE_LABELS,
    classify_provenance,
    resolve_provenance,
)
from memory.tags import retag_memory

from .. import deps


async def memory_detail(request: Request) -> HTMLResponse:
    """GET /memory/{key:path} — full memory detail view."""
    key = request.path_params["key"]
    data = deps.store.get(key)

    if data is None:
        template = request.app.state.templates.get_template("base.html")
        # Render a simple not-found within the base layout
        return HTMLResponse(
            template.render(
                request=request,
                current_page="memories",
            ).replace(
                "{% block content %}{% endblock %}",
                '<p class="empty-state">Memory not found.</p>',
            ),
            status_code=404,
        )

    # Parse namespace from key
    parts = key.split(":")
    namespace = parts[1] if len(parts) > 1 else "unknown"

    # Parse tags
    tags = []
    tags_raw = data.get("tags", "[]")
    try:
        tags = json.loads(tags_raw) if tags_raw else []
    except (json.JSONDecodeError, TypeError):
        pass

    # Parse abandoned approaches
    abandoned = []
    abandoned_raw = data.get("abandoned_approaches", "[]")
    try:
        abandoned = json.loads(abandoned_raw) if abandoned_raw else []
    except (json.JSONDecodeError, TypeError):
        pass

    # Parse contradictions
    contradictions = []
    contradictions_raw = data.get("contradictions", "[]")
    try:
        contradictions = json.loads(contradictions_raw) if contradictions_raw else []
    except (json.JSONDecodeError, TypeError):
        pass

    # Parse reinstate hints
    reinstate_hints = []
    hints_raw = data.get("reinstate_hints", "[]")
    try:
        reinstate_hints = json.loads(hints_raw) if hints_raw else []
    except (json.JSONDecodeError, TypeError):
        pass

    # Format timestamps
    def fmt_ts(raw):
        try:
            ts = float(raw)
            return time.strftime("%Y-%m-%d %H:%M:%S", time.localtime(ts))
        except (ValueError, TypeError):
            return "—"

    memory = {
        "key": key,
        "namespace": namespace,
        "content": data.get("content", ""),
        "state": data.get("state", "active"),
        "project": data.get("project") or data.get("project_name") or "",
        "tags": tags,
        "surface_score": data.get("surface_score", "1.0"),
        "experience_weight": data.get("experience_weight", "1.0"),
        "effort_score": data.get("effort_score"),
        "outcome": data.get("outcome"),
        "iterations": data.get("iterations"),
        "breakthrough": data.get("breakthrough"),
        "gotchas": data.get("gotchas"),
        "abandoned_approaches": abandoned,
        "contradictions": contradictions,
        "reinstate_hints": reinstate_hints,
        "deprioritised_reason": data.get("deprioritised_reason", ""),
        "source_url": data.get("source_url", ""),
        "feed_name": data.get("feed_name", ""),
        "licence": data.get("licence") or "",
        "licence_label": LICENCE_LABELS.get(data.get("licence") or "", "Not recorded"),
        "licence_note": data.get("licence_note") or "",
        "provenance": data.get("provenance") or "",
        "provenance_label": PROVENANCE_LABELS.get(data.get("provenance") or "", "Not recorded"),
        "recall_count": int(data.get("recall_count") or 0),
        "last_recalled": fmt_ts(data.get("last_recalled")) if data.get("last_recalled") else "Never",
        "created_at": fmt_ts(data.get("created_at")),
        "updated_at": fmt_ts(data.get("updated_at")),
    }

    template = request.app.state.templates.get_template("detail.html")
    content = template.render(
        request=request,
        memory=memory,
        tag_error=request.query_params.get("tag_error", ""),
        licence_error=request.query_params.get("licence_error", ""),
        licence_classes=LICENCE_CHOICES,
        provenance_error=request.query_params.get("provenance_error", ""),
        provenance_classes=PROVENANCE_CHOICES,
        current_page="memories",
    )
    return HTMLResponse(content)


async def memory_provenance(request: Request) -> RedirectResponse:
    """POST /memory/{key:path}/provenance — reclassify where a memory came from.

    Same engine as the set_provenance MCP tool: facts extracted from the
    memory follow it, and updated_at is left alone.
    """
    key = request.path_params["key"]
    if not is_classifiable_key(key) or deps.store.get_fields_multi([key], ("created_at",))[0] is None:
        return RedirectResponse(url=f"/memory/{key}", status_code=303)

    form = await request.form()
    try:
        classify_provenance(deps.store, [key], resolve_provenance(form.get("provenance", "")))
    except ValueError as exc:
        return RedirectResponse(
            url=f"/memory/{key}?provenance_error={quote(str(exc))}", status_code=303
        )
    return RedirectResponse(url=f"/memory/{key}", status_code=303)


async def memory_licence(request: Request) -> RedirectResponse:
    """POST /memory/{key:path}/licence — record a memory's redistribution rights.

    Same engine as the set_licence MCP tool (memory/licence.py
    classify_memories): the class must resolve, the note is optional and
    bounded, facts extracted from the memory follow it, and a
    reclassification never keeps the old class's note. Only the four
    licence-bearing namespaces are accepted — skills are derived, and
    nothing else under a mem:/meta: prefix is a memory.
    """
    key = request.path_params["key"]
    if not is_classifiable_key(key):
        return RedirectResponse(url=f"/memory/{key}", status_code=303)
    # get_fields_multi returns None when none of the projected fields exist,
    # not when the key is missing — so created_at (which every writer sets)
    # is what proves the record is there. An unstamped record must still be
    # classifiable; that is the whole point of the form.
    current = deps.store.get_fields_multi([key], ("created_at", "licence", "licence_note"))[0]
    if current is None:
        return RedirectResponse(url=f"/memory/{key}", status_code=303)

    form = await request.form()
    try:
        licence_class, derived_note = resolve_licence(form.get("licence", ""))
        note = note_for_reclassification(
            current.get("licence") or "", current.get("licence_note"),
            licence_class, validate_licence_note(form.get("licence_note", "")),
        ) or derived_note
        classify_memories(deps.store, [key], licence_class, note)
    except ValueError as exc:
        return RedirectResponse(
            url=f"/memory/{key}?licence_error={quote(str(exc))}", status_code=303
        )
    return RedirectResponse(url=f"/memory/{key}", status_code=303)


async def memory_retag(request: Request) -> RedirectResponse:
    """POST /memory/{key:path}/tags — replace a memory's tags from a comma-separated field."""
    key = request.path_params["key"]
    form = await request.form()
    raw = form.get("tags", "")
    tags = [t.strip() for t in raw.split(",") if t.strip()]

    try:
        retag_memory(deps.store, key, tags=tags)
    except ValueError as exc:
        return RedirectResponse(
            url=f"/memory/{key}?tag_error={quote(str(exc))}", status_code=303
        )
    return RedirectResponse(url=f"/memory/{key}", status_code=303)


routes = [
    # Must precede the greedy {key:path} detail route
    Route("/memory/{key:path}/tags", memory_retag, methods=["POST"]),
    Route("/memory/{key:path}/licence", memory_licence, methods=["POST"]),
    Route("/memory/{key:path}/provenance", memory_provenance, methods=["POST"]),
    Route("/memory/{key:path}", memory_detail),
]

File: crates/omnimem-settings/src/detail.rs
//! `/memory/{key}`: one memory in full, with its tag, licence and provenance
//! forms (`web_ui/routes/detail.py`). The forms go through the same engine
//! calls as the `retag`, `set_licence` and `set_provenance` tools, so facts
//! extracted from a memory follow its classification and `updated_at` is left
//! alone.

use std::collections::HashMap;

use axum::Form;
use axum::extract::{Path, Query, State};
use axum::http::StatusCode;
use axum::response::Response;
use chrono::{Local, TimeZone};
use minijinja::context;
use omnimem_engine::EngineError;
use omnimem_engine::classification::{
    is_classifiable_key, note_for_reclassification, resolve_licence, resolve_provenance,
    validate_licence_note,
};
use serde_json::{Value, json};

use crate::PanelState;
use crate::choices::{LICENCE_CHOICES, PROVENANCE_CHOICES, licence_label, provenance_label};
use crate::pages::{blocking, quote, see_other, starting};
use crate::render::page;

type FormData = HashMap<String, String>;

fn json_list(raw: Option<&String>) -> Value {
    raw.filter(|r| !r.is_empty())
        .and_then(|r| serde_json::from_str::<Value>(r).ok())
        .unwrap_or_else(|| json!([]))
}

fn fmt_ts(raw: Option<&String>) -> String {
    raw.and_then(|r| r.trim().parse::<f64>().ok())
        .and_then(|ts| Local.timestamp_opt(ts as i64, 0).single())
        .map_or_else(
            || "—".to_owned(),
            |t| t.format("%Y-%m-%d %H:%M:%S").to_string(),
        )
}

pub(crate) async fn handler(
    State(state): State<PanelState>,
    Path(key): Path<String>,
    Query(query): Query<HashMap<String, String>>,
) -> Response {
    let Some(engine) = state.engine() else {
        return starting(&state);
    };
    let lookup = key.clone();
    let data = match blocking(move || Ok(engine.store().get(&lookup)?)).await {
        Ok(data) => data,
        Err(failure) => return failure,
    };
    let Some(data) = data else {
        let mut response = page(
            state.templates(),
            "not_found.html",
            context! { current_page => "memories", message => "Memory not found." },
        );
        *response.status_mut() = StatusCode::NOT_FOUND;
        return response;
    };

    let text = |field: &str| data.get(field).cloned().unwrap_or_default();
    let present = |field: &str| data.get(field).cloned();
    let namespace = key.split(':').nth(1).unwrap_or("unknown").to_owned();
    let licence = text("licence");
    let provenance = text("provenance");
    let memory = json!({
        "key": key,
        "namespace": namespace,
        "content": text("content"),
        "state": data.get("state").cloned().unwrap_or_else(|| "active".to_owned()),
        "project": data.get("project").filter(|p| !p.is_empty()).or_else(|| data.get("project_name")).cloned().unwrap_or_default(),
        "tags": json_list(data.get("tags")),
        "surface_score": data.get("surface_score").cloned().unwrap_or_else(|| "1.0".to_owned()),
        "experience_weight": data.get("experience_weight").cloned().unwrap_or_else(|| "1.0".to_owned()),
        "effort_score": present("effort_score"),
        "outcome": present("outcome"),
        "iterations": present("iterations"),
        "breakthrough": present("breakthrough"),
        "lesson": present("lesson"),
        "gotchas": present("gotchas"),
        "abandoned_approaches": json_list(data.get("abandoned_approaches")),
        "contradictions": json_list(data.get("contradictions")),
        "reinstate_hints": json_list(data.get("reinstate_hints")),
        "deprioritised_reason": text("deprioritised_reason"),
        "source_url": text("source_url"),
        "feed_name": text("feed_name"),
        "licence_label": licence_label(&licence),
        "licence": licence,
        "licence_note": text("licence_note"),
        "provenance_label": provenance_label(&provenance),
        "provenance": provenance,
        "recall_count": data.get("recall_count").and_then(|c| c.trim().parse::<i64>().ok()).unwrap_or(0),
        "last_recalled": match data.get("last_recalled").filter(|l| !l.is_empty()) {
            Some(raw) => fmt_ts(Some(raw)),
            None => "Never".to_owned(),
        },
        "created_at": fmt_ts(data.get("created_at")),
        "updated_at": fmt_ts(data.get("updated_at")),
    });
    let param = |name: &str| query.get(name).cloned().unwrap_or_default();
    page(
        state.templates(),
        "detail.html",
        context! {
            memory,
            tag_error => param("tag_error"),
            licence_error => param("licence_error"),
            licence_classes => LICENCE_CHOICES,
            provenance_error => param("provenance_error"),
            provenance_classes => PROVENANCE_CHOICES,
            current_page => "memories",
        },
    )
}

/// Where a detail form lands: the memory, with the problem when there was one.
enum Outcome {
    Done,
    Refused(String),
}

async fn run_form(
    state: PanelState,
    key: String,
    error_param: &'static str,
    work: impl FnOnce(&omnimem_engine::Engine, &str) -> omnimem_engine::Result<Outcome> + Send + 'static,
) -> Response {
    let Some(engine) = state.engine() else {
        return starting(&state);
    };
    let worker_key = key.clone();
    let outcome = blocking(move || match work(&engine, &worker_key) {
        Err(EngineError::Invalid(message)) => Ok(Outcome::Refused(message)),
        other => other,
    })
    .await;
    match outcome {
        Ok(Outcome::Done) => see_other(&format!("/memory/{key}")),
        Ok(Outcome::Refused(message)) => {
            see_other(&format!("/memory/{key}?{error_param}={}", quote(&message)))
        }
        Err(failure) => failure,
    }
}

/// A tool result carrying `error` is a refusal, as a `ValueError` was.
fn refused(result: &Value) -> Option<String> {
    result["error"].as_str().map(str::to_owned)
}

pub(crate) async fn retag(
    State(state): State<PanelState>,
    Path(key): Path<String>,
    Form(form): Form<FormData>,
) -> Response {
    let tags: Vec<String> = form
        .get("tags")
        .map(|raw| {
            raw.split(',')
                .map(str::trim)
                .filter(|t| !t.is_empty())
                .map(str::to_owned)
                .collect()
        })
        .unwrap_or_default();
    run_form(state, key, "tag_error", move |engine, key| {
        engine.retag(key, Some(tags), None, None)?;
        Ok(Outcome::Done)
    })
    .await
}

pub(crate) async fn licence(
    State(state): State<PanelState>,
    Path(key): Path<String>,
    Form(form): Form<FormData>,
) -> Response {
    run_form(state, key, "licence_error", move |engine, key| {
        // created_at proves the record exists: every writer sets it, and an
        // unclassified record must still be classifiable here.
        if !is_classifiable_key(key) {
            return Ok(Outcome::Done);
        }
        let Some(current) = engine
            .store()
            .get_fields_multi(
                &[key.to_owned()],
                &["created_at", "licence", "licence_note"],
            )?
            .pop()
            .flatten()
        else {
            return Ok(Outcome::Done);
        };
        let field = |name: &str| form.get(name).map_or("", String::as_str);
        let (class, derived) = resolve_licence(field("licence"))?;
        let submitted = validate_licence_note(Some(field("licence_note")))?;
        // The form pre-fills the current note: switching class keeps it only
        // when the human changed it.
        let note = note_for_reclassification(
            current.get("licence").map_or("", String::as_str),
            current.get("licence_note").map(String::as_str),
            class,
            submitted,
        )
        .or_else(|| derived.map(str::to_owned));
        let result = engine.set_licence(class, Some(&[key.to_owned()]), None, note.as_deref())?;
        Ok(refused(&result).map_or(Outcome::Done, Outcome::Refused))
    })
    .await
}

pub(crate) async fn provenance(
    State(state): State<PanelState>,
    Path(key): Path<String>,
    Form(form): Form<FormData>,
) -> Response {
    run_form(state, key, "provenance_error", move |engine, key| {
        if !is_classifiable_key(key)
            || engine
                .store()
                .get_fields_multi(&[key.to_owned()], &["created_at"])?
                .pop()
                .flatten()
                .is_none()
        {
            return Ok(Outcome::Done);
        }
        let class = resolve_provenance(form.get("provenance").map_or("", String::as_str))?;
        let result = engine.set_provenance(class, &[key.to_owned()])?;
        Ok(refused(&result).map_or(Outcome::Done, Outcome::Refused))
    })
    .await
}
